content/blog/2019-10-23-register-allocation/index.md

```diff
@@ -0,0 +1,142 @@
+
+We also implemented a naïve register allocation technique, which assigns registers in a round-robin manner. This technique doesn't guarantee any efficiencies, but merely provides a mechanism to map to registers. This implementation presently does not support branch operations and therefore is only used for a subset of benchmarks as a sanity check.
+
+To gain more insight, we also manually allocated registers by graph coloring for some tests. This provides a reasonable ceiling to potential performance.
+
+### Counting the Number of Memory Access
+We modified the Bril interpreter `brili` to count the number of implicit memory access operations for those data not in registers. As discussed in the code generation section, we introduce a special rule, that is, variable names prefix with `r` are registers. Read and write operation on those registers do not require memory access. However, read and write on other variables all are counted as memory access. With this rule, we can do register allocation and perform evaluation on Bril without language breaking changes. For example:
+
+- `a: int = const 2;` has 1 memory access operations (store a)
+- `a: int = id b;` has 2 memory access operations (load b, store a)
+- `c:int = add a b;` has 3 memory access operations (load a, load b, store c)
+- `br given a b;` has 1 memory access operation (load given)
+- `print <N args>;` has N memory access operations, N being the number of arguments (N loads)
+
+As for accessing register values (the case where special variables represented registers will not be counted as a memory access):
+
+- `r1: int = const 2;` has 0 memory access operations
+- `a: int = id r2;` has 1 memory access operations (store a)
+- `c:int = add a r2;` has 2 memory access operations (load a, store c)
+
+We compare the total number of memory access operations for the same program with and without register allocation. 
+
+It was identified during evaluation that this implementation of counting doesn't support the `ret` instruction. We hope to fix this issue shortly. We have tested this interpreter version with some hand written register allocated code.
+
+### Simulated load and store counts
+
+We simulated the baseline and graph coloring implementation using Brili to test for two things,
+1. To make sure that the functionality hasn't been affected by our optimization
+2. To measure memory access operations as a proxy for performance
+
+The following tables provides the memory acccess count in the form of `loads/stores`. These tests are run assuming for 4 registers.
+
+Basline and graph coloring tests are an average from multiple simulation runs. This turned out to be useful for the graph coloring technique, as the register assignment can change.
```

Comment:
Why can the assignment change? Is some sort of iteration randomized or otherwise nondeterministic? If so, what is the source of the nondeterminism?